File: jerryscript/benchmarks/slre.js
var SCALE_FACTOR = 1;

var SLRE_NO_MATCH = -1;
var SLRE_UNEXPECTED_QUANTIFIER = -2;
var SLRE_UNBALANCED_BRACKETS = -3;
var SLRE_INTERNAL_ERROR = -4;
var SLRE_INVALID_CHARACTER_SET = -5;
var SLRE_INVALID_METACHARACTER = -6;
var SLRE_CAPS_ARRAY_TOO_SMALL = -7;
var SLRE_TOO_MANY_BRANCHES = -8;
var SLRE_TOO_MANY_BRACKETS = -9;

// OffsetString constructor
function OffsetString(data) {
    this.data = data;
    this.offset = 0;
    this.end = data.length;
}

OffsetString.prototype.toString = function () {
    return this.data.slice(this.offset, this.end);
};

OffsetString.prototype.len = function () {
    return this.end - this.offset;
};

OffsetString.prototype.charAt = function (index) {
    var str = this.toString();
    if (index < 0 || index >= str.length) return undefined;
    return str[index];
};

OffsetString.prototype.slice = function (start, stop) {
    var str = this.toString();
    if (stop === undefined) {
        return str.slice(start);
    }
    return str.slice(start, stop);
};

OffsetString.prototype.shallowCopy = function () {
    return new OffsetString(this.data);
};

OffsetString.prototype.advance = function (n) {
    this.offset += n;
};

OffsetString.prototype.advanceTo = function (offset) {
    this.offset += offset;
};

OffsetString.prototype.absSlice = function (start, length) {
    var res = this.shallowCopy();
    res.offset = start;
    res.end = start + length;
    return res;
};

OffsetString.prototype.relSlice = function (relStart, length) {
    return this.absSlice(this.offset + relStart, length);
};

// BracketPair constructor
function BracketPair(startOffset) {
    this.startOffset = startOffset;
    this.length = null;
    this.branchIndex = null;
    this.numBranches = null;
}

// Branch constructor
function Branch(bracketIndex, offset) {
    this.bracketIndex = bracketIndex;
    this.offset = offset;
}

// Capture constructor
function Capture(startOffset, length) {
    this.startOffset = (startOffset !== undefined) ? startOffset : null;
    this.length = (length !== undefined) ? length : null;
}

// RegexInfo constructor
function RegexInfo(captures) {
    this.brackets = [];
    this.branches = [];
    this.captures = captures;
    this.ignoreCase = false;
}

// Helper functions
var HEXDIGITS = "0123456789abcdefABCDEF";

function isHexDigit(c) {
    return HEXDIGITS.indexOf(c) !== -1;
}

function isSpace(ch) {
    return ch === ' ' || ch === '\t' || ch === '\n' || ch === '\r' || ch === '\f' || ch === '\v';
}

function isDigit(ch) {
    return ch >= '0' && ch <= '9';
}

function isMetacharacter(s) {
    return '^$().[]*+?|\\Ssd'.indexOf(s) !== -1;
}

function opLen(re) {
    if (re[0] === '\\') {
        if (re[1] === 'x') {
            return 4;
        }
        return 2;
    }
    return 1;
}

function setLen(re) {
    var i = 0;
    while (i < re.length && re[i] !== ']') {
        i += opLen(re.slice(i));
    }

    if (i <= re.length) {
        return i + 1;
    } else {
        throw new Error("Invalid character set");
    }
}

function getOpLen(re) {
    if (re[0] === '[') {
        return setLen(re.slice(1)) + 1;
    } else {
        return opLen(re);
    }
}

function isQuantifier(re) {
    return re[0] === '*' || re[0] === '+' || re[0] === '?';
}

function hextoi(s) {
    return parseInt(s, 16);
}

function matchOp(re, s, info) {
    var result = 0;

    switch (re[0]) {
        case '\\':
            switch (re[1]) {
                case 'S':
                    if (isSpace(s[0])) return SLRE_NO_MATCH;
                    result += 1;
                    break;
                case 's':
                    if (!isSpace(s[0])) return SLRE_NO_MATCH;
                    result += 1;
                    break;
                case 'd':
                    if (!isDigit(s[0])) return SLRE_NO_MATCH;
                    result += 1;
                    break;
                case 'x':
                    if (hextoi(re.slice(2, 4)) !== s.charCodeAt(0)) return SLRE_NO_MATCH;
                    result += 1;
                    break;
                default:
                    if (re[1] !== s[0]) return SLRE_NO_MATCH;
                    result += 1;
                    break;
            }
            break;
        case '|':
            return SLRE_INTERNAL_ERROR;
        case '$':
            return SLRE_NO_MATCH;
        case '.':
            result += 1;
            break;
        default:
            if (!s) {
                return SLRE_NO_MATCH;
            }
            if (info.ignoreCase) {
                if (re[0].toLowerCase() !== s[0].toLowerCase()) return SLRE_NO_MATCH;
            } else {
                if (re[0] !== s[0]) return SLRE_NO_MATCH;
            }
            result += 1;
            break;
    }
    return result;
}

function matchSet(re, reLen, s, info) {
    var i = 0;
    var result = -1;
    var invert = re[0] === '^';

    if (invert) {
        re = re.slice(1);
        reLen -= 1;
    }

    while (i <= reLen && re[i] !== ']' && result <= 0) {
        if (re[i] !== '-' && re[i + 1] === '-' && re[i + 2] !== ']' && i + 2 < re.length) {
            if (info.ignoreCase) {
                result = (re[i] <= s[0].toLowerCase() && s[0].toLowerCase() <= re[i + 2]) ? 1 : -1;
            } else {
                result = (re[i] <= s[0] && s[0] <= re[i + 2]) ? 1 : -1;
            }
            i += 3;
        } else {
            result = matchOp(re.slice(i), s, info);
            i += opLen(re.slice(i));
        }
    }

    var matched = result > 0;
    return (matched !== invert) ? 1 : -1;
}

function bar(re, s, info, bi) {
    var reOff = 0;
    var sOff = 0;

    while (reOff < re.len() && sOff <= s.len()) {
        var step;
        if (re.charAt(reOff) === '(') {
            step = info.brackets[bi + 1].length + 2;
        } else {
            step = getOpLen(re.slice(reOff));
        }

        if (step <= 0) {
            return SLRE_INVALID_CHARACTER_SET;
        }
        if (isQuantifier(re.toString())) {
            return SLRE_UNEXPECTED_QUANTIFIER;
        }

        if (reOff + step < re.len() && isQuantifier(re.charAt(reOff + step))) {
            if (re.charAt(reOff + step) === '?') {
                var res = bar(re.relSlice(reOff, step), s.relSlice(sOff, s.len() - sOff), info, bi);
                sOff += res > 0 ? res : 0;
                reOff += 1;
            } else if (re.charAt(reOff + step) === '+' || re.charAt(reOff + step) === '*') {
                var sOffInner = sOff;
                var nsOff = sOff;
                var nonGreedy = false;

                var afterQuant = reOff + step + 1;
                if (afterQuant < re.len() && re.charAt(afterQuant) === '?') {
                    nonGreedy = true;
                    afterQuant += 1;
                }

                var n1 = 1;
                var n2 = -1;
                while (n1 > 0) {
                    n1 = bar(
                        re.relSlice(reOff, step),
                        s.relSlice(sOffInner, s.len() - sOffInner),
                        info,
                        bi
                    );
                    if (n1 > 0) {
                        sOffInner += n1;
                    }
                    if (re.charAt(reOff + step) === '+' && n1 < 0) {
                        break;
                    }

                    if (afterQuant >= re.len()) {
                        nsOff = sOffInner;
                    } else {
                        n2 = bar(
                            re.relSlice(afterQuant, re.len() - afterQuant),
                            s.relSlice(sOffInner, s.len() - sOffInner),
                            info,
                            bi
                        );
                        if (n2 >= 0) {
                            nsOff = sOffInner + n2;
                        }
                    }

                    if (nsOff > sOff && nonGreedy) {
                        break;
                    }
                }

                if (n1 < 0 && re.charAt(reOff + step) === '*') {
                    n2 = bar(
                        re.relSlice(afterQuant, re.len() - afterQuant),
                        s.relSlice(sOffInner, s.len() - sOffInner),
                        info,
                        bi
                    );
                    if (n2 >= 0) {
                        nsOff = sOffInner + n2;
                    }
                }

                if (re.charAt(step) === '+' && nsOff === sOff) {
                    return SLRE_NO_MATCH;
                }

                if (nsOff === sOff && afterQuant < re.len() && n2 < 0) {
                    return SLRE_NO_MATCH;
                }

                return nsOff;
            }
            reOff += step;
            continue;
        }

        var reChar = re.charAt(reOff);

        switch (reChar) {
            case '[':
                var n = matchSet(re.slice(reOff + 1), re.len() - (reOff + 2), s.slice(sOff), info);
                if (n <= 0) return SLRE_NO_MATCH;
                sOff += n;
                break;
            case '(':
                var n = SLRE_NO_MATCH;
                bi += 1;
                if (bi >= info.brackets.length) {
                    return SLRE_INTERNAL_ERROR;
                }

                if (reOff + step >= re.len()) {
                    n = doh(re, s.relSlice(sOff, s.len() - sOff), info, bi);
                } else {
                    for (var j2 = 0; j2 <= s.len() - sOff; j2++) {
                        n = doh(re, s.relSlice(sOff, s.len() - (sOff + j2)), info, bi);
                        if (n < 0) continue;
                        var tmp = bar(
                            re.relSlice(reOff + step, re.len() - (reOff + step)),
                            s.relSlice(sOff + n, s.len() - (sOff + n)),
                            info,
                            bi
                        );
                        if (tmp >= 0) break;
                    }
                }

                if (n < 0) return n;

                while (info.captures.length < bi) {
                    info.captures.push(new Capture());
                }

                info.captures[bi - 1] = new Capture(s.offset + sOff, n);
                sOff += n;
                break;
            case '^':
                if (sOff !== 0) return SLRE_NO_MATCH;
                break;
            case '$':
                if (sOff !== s.len()) return SLRE_NO_MATCH;
                break;
            default:
                if (sOff >= s.len()) return SLRE_NO_MATCH;
                var n = matchOp(re.slice(reOff), s.slice(sOff), info);
                if (n <= 0) return n;
                sOff += n;
                break;
        }
        reOff += step;
    }
    return sOff;
}

function doh(re, s, info, bi) {
    var bp = info.brackets[bi];

    var pos;
    var length;
    var result = SLRE_NO_MATCH;

    var iterations = [0];
    for (var k = 1; k < bp.numBranches; k++) {
        iterations.push(k);
    }

    for (var idx = 0; idx < iterations.length; idx++) {
        var i = iterations[idx];
        if (i === 0) {
            pos = bp.startOffset;
        } else {
            pos = info.branches[bp.branchIndex + i - 1].offset + 1;
        }

        if (bp.numBranches === 0) {
            length = bp.length;
        } else if (bp.numBranches === i) {
            length = bp.startOffset + bp.length - pos;
        } else {
            length = info.branches[bp.branchIndex + i].offset - pos;
        }

        result = bar(re.absSlice(pos, length), s, info, bi);
        if (result > 0 || i >= bp.numBranches) {
            break;
        }
    }
    return result;
}

function baz(re, s, info) {
    var result = -1;
    var isAnchored = re.toString().indexOf('^') === 0;

    var sLen = s.toString().length;
    for (var i = 0; i < sLen; i++) {
        result = doh(re, s, info, 0);
        s.advance(1);
        if (result >= 0) {
            result += i;
            break;
        }
        if (isAnchored) {
            break;
        }
    }

    return result;
}

function setupBranchPoints(info) {
    var i, j, tmp;
    for (i = 0; i < info.branches.length; i++) {
        for (j = 0; j < info.branches.length; j++) {
            if (info.branches[i].bracketIndex > info.branches[j].bracketIndex) {
                tmp = info.branches[i];
                info.branches[i] = info.branches[j];
                info.branches[j] = tmp;
            }
        }
    }

    j = 0;

    for (i = 0; i < info.brackets.length; i++) {
        var bp = info.brackets[i];
        bp.numBranches = 0;
        bp.branchIndex = j;
        while (j < info.branches.length && info.branches[j].bracketIndex === i) {
            bp.numBranches += 1;
            j += 1;
        }
    }
}

function foo(re, s, info) {
    var depth = 0;

    var bracket = new BracketPair(re.offset);
    bracket.length = re.len();
    info.brackets.push(bracket);

    var iter = regexIterator(re);
    var iterResult = iter.next();
    while (!iterResult.done) {
        var seq = iterResult.value[0];
        var offset = iterResult.value[1];

        switch (seq[0]) {
            case '|':
                var currentBracketPair;
                if (info.brackets[info.brackets.length - 1].length === null) {
                    currentBracketPair = info.brackets.length;
                } else {
                    currentBracketPair = depth;
                }
                info.branches.push(new Branch(currentBracketPair, offset));
                break;
            case '\\':
                if (seq.length < 2 || (seq[1] === 'x'
                        && (seq.length < 4 || !allHexDigits(seq.slice(2, 4))))) {
                    return SLRE_INVALID_METACHARACTER;
                }
                break;
            case '(':
                depth += 1;
                info.brackets.push(new BracketPair(offset + 1));
                break;
            case ')':
                var idx;
                if (info.brackets[info.brackets.length - 1].length === null) {
                    idx = info.brackets.length - 1;
                } else {
                    idx = depth;
                }
                info.brackets[idx].length = offset - info.brackets[idx].startOffset;
                depth -= 1;
                if (depth < 0) {
                    return SLRE_UNBALANCED_BRACKETS;
                }
                if (re.data[offset - 1] === '(') {
                    return SLRE_NO_MATCH;
                }
                break;
            default:
                break;
        }

        iterResult = iter.next();
    }

    if (depth !== 0) {
        return SLRE_UNBALANCED_BRACKETS;
    }
    setupBranchPoints(info);

    return baz(re, s, info);
}

// Helper to replace [...seq.slice(2,4)].every(c => HEXDIGITS.includes(c))
function allHexDigits(s) {
    for (var i = 0; i < s.length; i++) {
        if (HEXDIGITS.indexOf(s[i]) === -1) {
            return false;
        }
    }
    return true;
}

// Replace generator function with a manual iterator
function regexIterator(reStr) {
    var iterOffset = reStr.offset;
    return {
        next: function () {
            if (iterOffset >= reStr.end) {
                return { done: true, value: undefined };
            }
            var iterStr = reStr.data.slice(iterOffset);
            var opLenVal = getOpLen(iterStr);
            var seq = iterStr.slice(0, opLenVal);
            var currentOffset = iterOffset;
            iterOffset += opLenVal;
            return { done: false, value: [seq, currentOffset] };
        }
    };
}

function slreMatch(regexp, sStr, caps) {
    var info = new RegexInfo(caps);
    var re = new OffsetString(regexp);
    var s = new OffsetString(sStr);

    if (re.toString().indexOf('(?i)') === 0) {
        info.ignoreCase = true;
        re.advance(4);
    }

    return foo(re, s, info);
}

function initialize_benchmark() {

}

function benchmark() {
    initialize_benchmark()
    var res = benchmarkBody(SCALE_FACTOR);
    return verifyBenchmark(res);
}

function benchmarkBody(lsf) {
    var text = "abbbababaabccababcacbcbcbabbabcbabcabcbbcbbac";
    var regexes = ["(ab)+", "(b.+)+", "a[ab]*", "([ab^c][ab^c])+"];

    var ret = 0;
    for (var i = 0; i < lsf; i++) {
        ret = 0;
        var captures = [];
        for (var r = 0; r < regexes.length; r++) {
            ret += slreMatch(regexes[r], text, captures);
        }
    }

    return ret;
}

function verifyBenchmark(r) {
    return 102 === r;
}

benchmark();

File: micropython/benchmarks/slre.py
# import micropython

SCALE_FACTOR = 1


class ReturnCodes(int):
    SLRE_NO_MATCH = -1
    SLRE_UNEXPECTED_QUANTIFIER = -2
    SLRE_UNBALANCED_BRACKETS = -3
    SLRE_INTERNAL_ERROR = -4
    SLRE_INVALID_CHARACTER_SET = -5
    SLRE_INVALID_METACHARACTER = -6
    SLRE_CAPS_ARRAY_TOO_SMALL = -7
    SLRE_TOO_MANY_BRANCHES = -8
    SLRE_TOO_MANY_BRACKETS = -9


class OffsetString:
    def __init__(self, data):
        self.data = data
        self.end = len(data)
        self.offset = 0

    def __str__(self):
        return self.data[self.offset:self.end]

    def __len__(self):
        assert self.offset <= self.end
        return self.end - self.offset

    def __getitem__(self, key):
        return str(self)[key]

    def shallow_copy(self):
        return OffsetString(self.data)

    def advance(self, n):
        self.offset += n

    def advance_to(self, offset):
        self.offset += offset

    def abs_slice(self, start, length):
        res = self.shallow_copy()
        res.offset = start
        res.end = start + length
        return res

    def rel_slice(self, rel_start, length):
        return self.abs_slice(self.offset + rel_start, length)


def regex_iterate(re_str):
    # Build and return a list of (offset, op_length) tuples for the regex
    # string. A proper iterator would be prefed here but the micropython
    # implementation on arm crashes when using an iterator here. I did not
    # have time to replicate this issue yet
    result = []
    iter_offset = re_str.offset
    while iter_offset < re_str.end:
        iter_str = re_str.data[iter_offset:]
        op_length = get_op_len(iter_str)
        result.append((iter_offset, op_length))
        iter_offset += op_length
    return result


class BracketPair:
    def __init__(self, start_offset):
        self.start_offset = start_offset
        self.length = None
        self.branch_index = None
        self.num_branches = None


class Branch:
    def __init__(self, bracket_index, offset):
        self.bracket_index = bracket_index
        self.offset = offset


class Capture:
    def __init__(self, start_offset=None, length=None):
        self.start_offset = start_offset
        self.length = length


class RegexInfo:
    def __init__(self, captures):
        self.brackets = []
        self.branches = []
        self.captures = captures
        self.ignore_case = False


def is_metacharacter(s):
    return s in '^$().[]*+?|\\Ssd'


def op_len(re):
    if re[0] == '\\':
        if re[1] == 'x':
            return 4
        return 2
    return 1


def set_len(re):
    i = 0
    while i < len(re) and re[i] != ']':
        i += op_len(re[i])

    if i <= len(re):
        return i + 1
    else:
        # This only happens when a regex ends with \x but does not supply
        # two characters afterward.
        raise ValueError


def get_op_len(re):
    if re[0] == '[':
        return set_len(re[1:]) + 1
    else:
        return op_len(re)


def is_quantifier(re):
    return re[0] in ['*', '+', '?']


def hextoi(s):
    return int(s, 16)


def match_op(re, s, info):
    result = 0
    if re[0] == '\\':
        # Metacharacters
        if re[1] == 'S':
            if s[0].isspace():
                return ReturnCodes.SLRE_NO_MATCH
            result += 1
        elif re[1] == 's':
            if not s[0].isspace():
                return ReturnCodes.SLRE_NO_MATCH
            result += 1
        elif re[1] == 'd':
            if not s[0].isdigit():
                return ReturnCodes.SLRE_NO_MATCH
            result += 1
        elif re[1] == 'x':
            # \xHH where HH is hexadecimal byte representation
            if hextoi(re[2]) != s[0]:
                return ReturnCodes.SLRE_NO_MATCH
            result += 1
        else:
            # Valid metacharacter check is done in bar()
            if re[1] != s[0]:
                return ReturnCodes.SLRE_NO_MATCH
            result += 1
    elif re[0] == '|':
        return ReturnCodes.SLRE_INTERNAL_ERROR
    elif re[0] == '$':
        return ReturnCodes.SLRE_NO_MATCH
    elif re[0] == '.':
        result += 1
    else:
        if not s:
            # String is empty
            return ReturnCodes.SLRE_NO_MATCH
        if info.ignore_case:
            if re[0].lower() != s[0].lower():
                return ReturnCodes.SLRE_NO_MATCH
        else:
            if re[0] != s[0]:
                return ReturnCodes.SLRE_NO_MATCH
        result += 1
    return result


def match_set(re, re_len, s, info):
    i = 0
    result = -1
    invert = re[0] == '^'

    if invert:
        re += 1
        re_len -= 1

    while i <= re_len and re[i] != ']' and result <= 0:
        if re[i] != '-' and re[i + 1] == '-' and re[i + 2] != ']' and i + 2 > len(re):
            if info.ignore_case:
                result = re[i] <= s[0] <= re[i + 2]
            else:
                result = re[i] <= s[0].lower() <= re[i + 2]
            i += 3
        else:
            result = match_op(re[i:], s, info)
            i += op_len(re[i:])

    matched = result > 0
    # we want to return 1 if either inverted is true and matched not True
    # or the other way around.
    # Conversely, If both are true or not true we return -1
    return 1 if matched ^ invert else -1


def bar(re, s, info, bi):
    re_off = 0
    s_off = 0

    while re_off < len(re) and s_off <= len(s):
        if re[re_off] == '(':
            assert info.brackets[bi + 1].length is not None
            step = info.brackets[bi + 1].length + 2
        else:
            step = get_op_len(re[re_off:])
        if step <= 0:
            return ReturnCodes.SLRE_INVALID_CHARACTER_SET
        if is_quantifier(str(re)):
            return ReturnCodes.SLRE_UNEXPECTED_QUANTIFIER

        if re_off + step < len(re) and is_quantifier(re[re_off + step]):
            if re[re_off + step] == '?':
                res = bar(re.rel_slice(re_off, step), s.rel_slice(s_off, len(s) - s_off), info, bi)
                s_off += res if res > 0 else 0
                re_off += 1
            elif re[re_off + step] in ['+', '*']:
                s_off_inner = s_off
                ns_off = s_off
                non_greedy = False

                after_quant = re_off + step + 1
                if after_quant < len(re) and re[after_quant] == '?':
                    non_greedy = True
                    after_quant += 1

                n1 = 1
                n2 = -1
                while n1 > 0:
                    n1 = bar(
                        re.rel_slice(re_off, step),
                        s.rel_slice(s_off_inner, len(s) - s_off_inner),
                        info,
                        bi
                    )
                    if n1 > 0:
                        s_off_inner += n1
                    if re[re_off + step] == '+' and n1 < 0:
                        break

                    if after_quant >= len(re):
                        ns_off = s_off_inner
                    else:
                        n2 = bar(
                            re.rel_slice(after_quant, len(re) - after_quant),
                            s.rel_slice(s_off_inner, len(s) - s_off_inner),
                            info,
                            bi
                        )
                        if n2 >= 0:
                            ns_off = s_off_inner + n2

                    if ns_off > s_off and non_greedy:
                        break

                if n1 < 0 and re[re_off + step] == '*':
                    n2 = bar(
                        re.rel_slice(after_quant, len(re) - after_quant),
                        s.rel_slice(s_off_inner, len(s) - s_off_inner),
                        info,
                        bi
                    )
                    if n2 >= 0:
                        ns_off = s_off_inner + n2

                if re[step] == '+' and ns_off == s_off:
                    return ReturnCodes.SLRE_NO_MATCH

                if ns_off == s_off and after_quant < len(re) and n2 < 0:
                    return ReturnCodes.SLRE_NO_MATCH

                return ns_off
            re_off += step
            continue

        if re[re_off] == '[':
            n = match_set(re[re_off + 1:], len(re) - (re_off + 2), s[s_off:], info)
            if n <= 0:
                return ReturnCodes.SLRE_NO_MATCH
            s_off += n
        elif re[re_off] == '(':
            n = ReturnCodes.SLRE_NO_MATCH
            bi += 1
            if bi >= len(info.brackets):
                return ReturnCodes.SLRE_INTERNAL_ERROR

            if re_off + step >= len(re):
                n = doh(re, s.rel_slice(s_off, len(s) - s_off), info, bi)
            else:
                for j2 in range(0, len(s) - s_off + 1):
                    n = doh(re, s.rel_slice(s_off, len(s) - (s_off + j2)), info, bi)
                    if n < 0:
                        continue
                    tmp = bar(
                        re.rel_slice(re_off + step, len(re) - (re_off + step)),
                        s.rel_slice(s_off + n, len(s) - (s_off + n)),
                        info,
                        bi
                    )
                    if tmp >= 0:
                        break

            if n < 0:
                return n
            while len(info.captures) < bi:
                info.captures.append(Capture())

            info.captures[bi - 1] = Capture(s.offset + s_off, n)
            s_off += n
        elif re[re_off] == '^':
            if s_off != 0:
                return ReturnCodes.SLRE_NO_MATCH
        elif re[re_off] == '$':
            if s_off != len(s):
                return ReturnCodes.SLRE_NO_MATCH
        else:
            if s_off >= len(s):
                return ReturnCodes.SLRE_NO_MATCH
            n = match_op(re[re_off:], s[s_off], info)
            if n <= 0:
                return n
            s_off += n
        re_off += step
    return s_off


# Process branch points
def doh(re, s, info, bi):
    bp = info.brackets[bi]

    # position of the branch we are matching in the regex str
    pos = 0
    # length of the Branch
    length = 0

    for i in [0] + list(range(1, bp.num_branches)):
        if i == 0:
            pos = bp.start_offset
        else:
            pos = info.branches[bp.branch_index + i - 1].offset + 1

        if bp.num_branches == 0:
            length = bp.length
        elif bp.num_branches == i:
            length = bp.start_offset + bp.length - pos
        else:
            length = info.branches[bp.branch_index + i].offset - pos

        result = bar(re.abs_slice(pos, length), s, info, bi)
        if result > 0 or i >= bp.num_branches:
            break
    return result


def baz(re, s, info):
    result = -1
    is_anchored = str(re).startswith('^')

    for i in range(len(str(s))):
        result = doh(re, s, info, 0)
        s.advance(1)
        if result >= 0:
            result += i
            break
        if is_anchored:
            break

    return result


def setup_branch_points(info):
    for i in range(len(info.branches)):
        for j in range(len(info.branches)):
            if info.branches[i].bracket_index > info.branches[j].bracket_index:
                tmp = info.branches[i]
                info.branches[i] = info.branches[j]
                info.branches[j] = tmp

    j = 0

    for i, bp in enumerate(info.brackets):
        bp.num_branches = 0
        bp.branch_index = j
        while j < len(info.branches) and info.branches[j].bracket_index == i:
            bp.num_branches += 1
            j += 1


def foo(re, s, info):
    depth = 0
    hexdigits = "0123456789abcdefABCDEF"

    # First bracket captures everything
    bracket = BracketPair(re.offset)
    bracket.length = len(re)
    info.brackets.append(bracket)

    # Build the list of (offset, op_length) pairs up front
    items = regex_iterate(re)

    # Make a single pass over regex string, memorize brackets and branches
    for (offset, op_length) in items:
        # Build the operator string from the offset and precomputed length
        seq = re.data[offset:offset + op_length]

        if seq[0] == '|':
            if info.brackets[-1].length is None:
                # We are currently in a bracket
                current_bracket_pair = len(info.brackets)
            else:
                # We have left all brackets
                current_bracket_pair = depth
            info.branches.append(Branch(current_bracket_pair, offset))
        elif seq[0] == '\\':
            # this is the case where the iterator raises a ValueError
            if len(seq) < 2 or seq[1] == 'x' \
                    and (len(seq) < 4 or not all(c in hexdigits for c in seq[2:4])):
                return ReturnCodes.SLRE_INVALID_METACHARACTER
        elif seq[0] == '(':
            depth += 1
            info.brackets.append(BracketPair(offset + 1))
        elif seq[0] == ')':
            # I think this (and the original) does not work for arbitrarily
            # nested capture groups.
            # for example (  )  (  (  )  )  (  (  )  )
            #             ^  ^  ^  ^  ^  ^  ^  ^  ^  ^
            #             A     B  C        D  E
            #     depth = 1  0  1  2  1  0  1  2  1  0
            #
            # info.brackets = [A, B, C, D, E]
            #                  0  1  2  3  4
            #
            # When we close a bracket we check if the top most bracket pair (bp)
            # is open (length is None) and if it is we close it.
            # If it is not we instead use the depth as an index to close it.
            # This means for A, C, and E we use the top most bp.
            # For B, we use the depth before closing as an index: 1 => B.
            # For D we also use the depth as an index: 1 => B
            # Here we use the wrong bp pair. We would need a per depth count of
            # closed brackets.
            i = len(info.brackets) - 1 if info.brackets[-1].length is None else depth
            info.brackets[i].length = offset - info.brackets[i].start_offset
            depth -= 1
            if depth < 0:
                return ReturnCodes.SLRE_UNBALANCED_BRACKETS
            if re[offset - 1] == '(':
                # We have a () group
                return ReturnCodes.SLRE_NO_MATCH
        else:
            pass

    if depth != 0:
        return ReturnCodes.SLRE_UNBALANCED_BRACKETS
    setup_branch_points(info)

    return baz(re, s, info)


def slre_match(regexp, s_str, caps):
    info = RegexInfo(caps)
    re = OffsetString(regexp)
    s = OffsetString(s_str)

    if str(re).startswith('(?i)'):
        info.ingnore_case = True
        re.advance(4)

    return foo(re, s, info)


def initialize_benchmark():
    pass


def benchmark():
    initialize_benchmark()
    res = benchmark_body(SCALE_FACTOR)
    # print("")
    # micropython.mem_info()
    return verify_benchmark(res)


def benchmark_body(lsf):
    text = "abbbababaabccababcacbcbcbabbabcbabcabcbbcbbac"
    regexes = ["(ab)+", "(b.+)+", "a[ab]*", "([ab^c][ab^c])+"]

    ret = 0
    for _ in range(lsf):
        captures = []
        for r in regexes:
            ret += slre_match(r, text, captures)

    return ret


def verify_benchmark(r):
    return 102 * SCALE_FACTOR == r
